server/base/src/main/java/org/apache/accumulo/server/manager/state/ManagerTabletInfoIterator.java

```diff
@@ -0,0 +1,434 @@
+    String managerStateOptionValue = options.get(MANAGER_STATE_OPTION);
+    try {
+      managerState = ManagerState.valueOf(managerStateOptionValue);
+    } catch (RuntimeException ex) {
+      if (managerStateOptionValue != null) {
+        LOG.error("Unable to decode managerState {}", managerStateOptionValue);
+      }
+    }
+    Set<TServerInstance> shuttingDown = parseServers(options.get(SHUTTING_DOWN_OPTION));
+    if (current != null && shuttingDown != null) {
+      current.removeAll(shuttingDown);
+    }
+  }
+
+  @Override
+  public Key getTopKey() {
+    return topKey;
+  }
+
+  @Override
+  public Value getTopValue() {
+    return topValue;
+  }
+
+  @Override
+  public boolean hasTop() {
+    return topKey != null && topValue != null;
+  }
+
+  @Override
+  protected void consume() throws IOException {
+    topKey = null;
+    topValue = null;
+
+    final Set<ManagementAction> reasonsToReturnThisTablet = new HashSet<>();
+    while (getSource().hasTop()) {
+      final Key k = getSource().getTopKey();
+      final Value v = getSource().getTopValue();
+      final SortedMap<Key,Value> decodedRow = WholeRowIterator.decodeRow(k, v);
+      final TabletMetadata tm = TabletMetadata.convertRow(decodedRow.entrySet().iterator(),
+          ManagerTabletInfo.CONFIGURED_COLUMNS, false, true);
+
+      LOG.trace("Evaluating extent: {}", tm);
+      reasonsToReturnThisTablet.clear();
+      computeTabletManagementActions(tm, reasonsToReturnThisTablet);
+      if (!reasonsToReturnThisTablet.isEmpty()) {
+        // If we simply returned here, then the client would get the encoded K,V
+        // from the WholeRowIterator. However, it would not know the reason(s) why
+        // it was returned. Insert a K,V pair to represent the reasons. The client
+        // can pull this K,V pair from the results by looking at the colf.
+        ManagerTabletInfo.addActions(decodedRow, reasonsToReturnThisTablet);
+        topKey = decodedRow.firstKey();
+        topValue = WholeRowIterator.encodeRow(new ArrayList<>(decodedRow.keySet()),
+            new ArrayList<>(decodedRow.values()));
+        LOG.trace("Returning extent {} with reasons: {}", tm.getExtent(),
+            reasonsToReturnThisTablet);
+        return;
+      }
+
+      LOG.trace("No reason to return extent {}, continuing", tm.getExtent());
+      getSource().next();
+    }
+  }
+
+  /**
+   * Evaluates whether or not this Tablet should be returned so that it can be acted upon by the
+   * Manager
+   */
+  private void computeTabletManagementActions(final TabletMetadata tm,
+      final Set<ManagementAction> reasonsToReturnThisTablet) {
+
+    if (onlineTables == null || current == null || managerState != ManagerState.NORMAL
+        || tm.isFutureAndCurrentLocationSet()) {
+      // no need to check everything, we are in a known state where we want to return everything.
+      reasonsToReturnThisTablet.add(ManagementAction.BAD_STATE);
+      return;
+    }
```

Comment:
With the changes to assume we always have servers and tables could do the following.   Also looked at the manager state it has some things related to shutdown.  I think the old iterator just returned everything when it was not normal, will that work with these new changes?

```suggestion
    if(managerState != ManagerState.NORMAL){
       // when manager is in the process of starting up or shutting down return everything.
       reasonsToReturnThisTablet.add(ManagementAction.NEEDS_LOCATION_UPDATE);
       return;
    }
    if ( tm.isFutureAndCurrentLocationSet()) {
      // no need to check everything, we are in a known state where we want to return everything.
      reasonsToReturnThisTablet.add(ManagementAction.BAD_STATE);
      return;
    }
```

Reply:
Implemented in 9153726